File: frontend/src/types/position.ts
/**
 * 持仓类型定义
 * PRD 4.18 分策略持仓管理
 */

/**
 * 策略持仓
 */
export interface StrategyPosition {
  position_id: string
  user_id: string
  account_id: string
  strategy_id: string | null
  strategy_name: string | null

  symbol: string
  quantity: number
  avg_cost: number
  current_price: number
  market_value: number

  unrealized_pnl: number
  unrealized_pnl_pct: number
  realized_pnl: number

  stop_loss?: number
  take_profit?: number

  beta?: number
  volatility?: number

  created_at: string
  updated_at: string
}

/**
 * 持仓来源
 */
export interface PositionSource {
  strategy_id: string | null
  strategy_name: string
  quantity: number
  avg_cost: number
  pnl: number
  pnl_pct: number
}

/**
 * 同股票汇总持仓
 */
export interface ConsolidatedPosition {
  symbol: string
  total_quantity: number
  weighted_avg_cost: number
  current_price: number
  total_market_value: number
  total_unrealized_pnl: number
  total_unrealized_pnl_pct: number
  sources: PositionSource[]
  concentration_pct: number
}

/**
 * 持仓分组
 */
export interface PositionGroup {
  strategy_id: string | null
  strategy_name: string
  positions: StrategyPosition[]
  total_market_value: number
  total_unrealized_pnl: number
  total_unrealized_pnl_pct: number
  position_count: number
}

/**
 * 账户持仓汇总
 */
export interface AccountPositionSummary {
  account_id: string
  total_market_value: number
  total_cash: number
  total_equity: number
  total_unrealized_pnl: number
  total_unrealized_pnl_pct: number
  groups: PositionGroup[]
  consolidated: ConsolidatedPosition[]
  concentration_warnings: string[]
  portfolio_beta?: number
  updated_at: string
}

/**
 * 卖出请求
 */
export interface SellPositionRequest {
  position_id: string
  quantity: number
  order_type?: 'market' | 'limit'
  limit_price?: number
}

/**
 * 卖出响应
 */
export interface SellPositionResponse {
  success: boolean
  order_id?: string
  message: string
}

/**
 * 视图模式
 */
export type PositionViewMode = 'strategy' | 'symbol'

/**
 * 格式化盈亏
 */
export function formatPnL(pnl: number, showSign = true): string {
  const sign = showSign && pnl >= 0 ? '+' : ''
  return `${sign}$${pnl.toLocaleString(undefined, {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  })}`
}

/**
 * 格式化盈亏百分比
 */
export function formatPnLPct(pct: number, showSign = true): string {
  const sign = showSign && pct >= 0 ? '+' : ''
  return `${sign}${pct.toFixed(2)}%`
}

/**
 * 格式化货币
 */
export function formatCurrency(value: number): string {
  return `$${value.toLocaleString(undefined, {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  })}`
}

/**
 * 获取盈亏颜色类
 */
export function getPnLColorClass(pnl: number): string {
  return pnl >= 0 ? 'text-green-400' : 'text-red-400'
}

File: backend/app/schemas/position.py
"""
分策略持仓 Schema
PRD 4.18 分策略持仓管理
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class StrategyPosition(BaseModel):
    """策略持仓 (逻辑隔离)"""
    position_id: str
    user_id: str
    account_id: str
    strategy_id: Optional[str] = None  # None = 手动交易
    strategy_name: Optional[str] = None

    symbol: str
    quantity: int
    avg_cost: float
    current_price: float
    market_value: float

    unrealized_pnl: float
    unrealized_pnl_pct: float
    realized_pnl: float = 0  # 已实现盈亏

    # 止盈止损
    stop_loss: Optional[float] = None
    take_profit: Optional[float] = None

    # 风险指标
    beta: Optional[float] = None
    volatility: Optional[float] = None

    created_at: datetime
    updated_at: datetime


class PositionSource(BaseModel):
    """持仓来源"""
    strategy_id: Optional[str] = None
    strategy_name: str
    quantity: int
    avg_cost: float
    pnl: float
    pnl_pct: float


class ConsolidatedPosition(BaseModel):
    """同股票汇总持仓"""
    symbol: str
    total_quantity: int
    weighted_avg_cost: float
    current_price: float
    total_market_value: float
    total_unrealized_pnl: float
    total_unrealized_pnl_pct: float

    # 来源策略
    sources: list[PositionSource]

    # 集中度
    concentration_pct: float = 0  # 占账户比例


class PositionGroup(BaseModel):
    """持仓分组 (按策略)"""
    strategy_id: Optional[str] = None
    strategy_name: str  # "手动交易" 或策略名
    positions: list[StrategyPosition]
    total_market_value: float
    total_unrealized_pnl: float
    total_unrealized_pnl_pct: float
    position_count: int = 0


class AccountPositionSummary(BaseModel):
    """账户持仓汇总"""
    account_id: str
    total_market_value: float
    total_cash: float
    total_equity: float
    total_unrealized_pnl: float
    total_unrealized_pnl_pct: float

    # 按策略分组
    groups: list[PositionGroup]

    # 同股票汇总视图
    consolidated: list[ConsolidatedPosition]

    # 风险指标
    concentration_warnings: list[str] = []  # 集中度警告
    portfolio_beta: Optional[float] = None

    updated_at: datetime


class SellPositionRequest(BaseModel):
    """卖出持仓请求"""
    position_id: str
    quantity: int = Field(..., ge=1)
    order_type: str = Field(default="market")
    limit_price: Optional[float] = None


class SellPositionResponse(BaseModel):
    """卖出持仓响应"""
    success: bool
    order_id: Optional[str] = None
    message: str


class UpdateStopLossRequest(BaseModel):
    """更新止损请求"""
    stop_loss: Optional[float] = Field(None, ge=0)
    take_profit: Optional[float] = Field(None, ge=0)


class PositionRiskMetrics(BaseModel):
    """持仓风险指标"""
    symbol: str
    beta: float
    volatility: float
    var_95: float  # 95% VaR
    max_loss: float
    correlation_to_spy: float
